bokehjs/src/coffee/models/tools/edit/point_draw_tool.ts

```diff
@@ -0,0 +1,138 @@
+  model: PointDrawTool
+  _basepoint: [number, number] | null
+
+  _tap(e: BkEv): void {
+    const append = e.srcEvent.shiftKey != null ? e.srcEvent.shiftKey : false;
+    const renderers = this._select_event(e, append, this.model.renderers);
+    if (renderers.length || !this.model.add) {
+      return
+    }
+
+    const renderer = this.model.renderers[0];
+    const point = this._map_drag(e.bokeh.sx, e.bokeh.sy, renderer);
+    if (point == null) {
+      return;
+    }
+    const glyph = renderer.glyph
+    const ds = renderer.data_source;
+    const [xkey, ykey] = [glyph.x.field, glyph.y.field];
+    let xs = ds.data[xkey];
+    let ys = ds.data[ykey];
+
+    // Convert typed arrays to regular arrays for easy manipulation
+    if ((xs.push == null)) {
+      ds.data[xkey] = (xs = Array.prototype.slice.call(xs));
+    }
+    if ((ys.push == null)) {
+      ds.data[ykey] = (ys = Array.prototype.slice.call(ys));
+    }
+
+    // Add x- and y-values into column arrays
+    const [x, y] = point;
+    xs.push(x);
+    ys.push(y);
+    this._pad_empty_columns(ds, [xkey, ykey]);
+
+    ds.change.emit(undefined);
+    ds.properties.data.change.emit(undefined);
+  }
+
+  _keyup(e: BkEv): void {
+    if (e.shiftKey && (e.keyCode === Keys.Backspace) && this.model.active) {
+      for (const renderer of this.model.renderers) {
+        this._delete_selected(renderer);
+      }
+    }
+  }
+
+  _pan_start(e: BkEv): void {
+    if (this.model.drag) {
+      const append = e.srcEvent.shiftKey != null ? e.srcEvent.shiftKey : false;
+      this._select_event(e, append, this.model.renderers);
+      this._basepoint = [e.bokeh.sx, e.bokeh.sy];
+    }
+  }
+
+  _pan(e: BkEv): void {
+    if (!this.model.drag || this._basepoint == null) {
+      return;
+    }
+    // If a Point is selected drag it
+    const [bx, by] = this._basepoint;
+    for (const renderer of this.model.renderers) {
+      const basepoint = this._map_drag(bx, by, renderer);
+      const point = this._map_drag(e.bokeh.sx, e.bokeh.sy, renderer);
+      if (point == null || basepoint == null) {
+        continue;
+      }
+
+      const [x, y] = point;
+      const [px, py] = basepoint;
+      const [dx, dy] = [x-px, y-py];
+      const glyph = renderer.glyph;
+      const ds = renderer.data_source;
+      const [xkey, ykey] = [glyph.x.field, glyph.y.field];
+      for (const index of ds.selected['1d'].indices) {
+        ds.data[xkey][index] += dx;
+        ds.data[ykey][index] += dy;
+      }
+      ds.change.emit(undefined);
+      ds.properties.data.change.emit(undefined);
+    }
+    this._basepoint = [e.bokeh.sx, e.bokeh.sy];
+  }
+
+  _pan_end(_e: BkEv): void {
+    if (!this.model.drag) { return; }
+    for (const renderer of this.model.renderers) {
+      renderer.data_source.selected['1d'].indices = [];
+    }
+    this._basepoint = null;
+  }
+}
+
+
+export class PointDrawTool extends EditTool {
+  add: boolean
+  drag: boolean
+  renderers: GlyphRenderer[] & Array<HasXYCDS>
```

Comment:
@mattpap Just wanted to check with you whether this is the proper way to handle this. Each tool can only work with renderers which render certain glyph types and use a ColumnDataSource so I've declared an intersection interface to type that (note that as I discussed with Bryan, Python will also validate this during serialization). This gets rid of all the TypeScript errors warning about ``.data`` not existing on a ``DataSource``. The only errors then remaining are these:
```
[13:14:39] src/coffee/models/tools/edit/point_draw_tool.ts(43,33): error TS2339: Property 'x' does not exist on type 'Glyph & XYGlyph'.
```

This is because the x and y properties are not yet typed on the glyphs. Advice on handling that would be very helpful too.

Reply:
That's a lot to unpack here. Some errors went away because there's a syntax error just before definition of `HasCDS`. The type should be `(GlyphRenderer & HasXYCDS)[]`, though maybe they are equivalent (I need check the rules for intersection types). I will be able to say more when the basics are fixed.

Reply:
Thanks, fixed the syntax now, the errors didn't come back. That said those definitely are not equivalent and those checks are only passing because my type definition gets interpreted as "a list of GlyphRenderers **or** a list of things that have a XY CDS", rather than your more correct suggestion that it should be "a list of GlyphRenderers that have an XY CDS".

I'll have to do some more reading.